### PR TITLE
Workaround for memory annotation problem (SCP-2143)

### PR DESCRIPTION
**Pull request by kwxm**
This is a quick workaround for the memory annotation problem that was affecting the costing results.  There's an issue for a fuller reworking at SCP-2139.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by kwxm** [comment]
The builtin cost estimates were being inflated by this. Here are the builtin times for the Marlowe zerocoupon1 validation before 
```
GreaterThanInteger       360736       10000
LessThanEqInteger       1329778      130000
IfThenElse              1980000     1980000
MultiplyInteger        18481948  1567940000
EqByteString           15260669      710000
SubtractInteger         2038790  1567980000
EqInteger              49446897     1080000
AddInteger              7167730  6271800000
LessThanInteger         1464006       50000
```
and after
```
GreaterThanInteger        21062       10000
LessThanEqInteger        272662      130000
IfThenElse              1980000     1980000
MultiplyInteger           38958       40000
EqByteString            1486734      710000
SubtractInteger           84484       80000
EqInteger               2256012     1080000
AddInteger               211850      200000
LessThanInteger          105310       50000
```
(the units are ms*10000, just to be confusing).  The total execution time was inflated by a factor of 15...

**Comment by kwxm** [comment]
I wondered if this was slowing things down, so I checked on the server:

```
bench-compare 16-Apr-master-validation 16-Apr-memory-workaround-validation 
crowdfunding/1           2.923 ms	2.131 ms    -27.1%
crowdfunding/2           2.927 ms	2.141 ms    -26.9%
crowdfunding/3           2.919 ms	2.143 ms    -26.6%
crowdfunding/4           1.794 ms	1.104 ms    -38.5%
crowdfunding/5           1.792 ms	1.102 ms    -38.5%
future/1                 1.749 ms	1.250 ms    -28.5%
future/2                 3.391 ms	2.544 ms    -25.0%
future/3                 3.392 ms	2.543 ms    -25.0%
future/4                 4.780 ms	3.114 ms    -34.9%
future/5                 5.394 ms	3.838 ms    -28.8%
future/6                 4.578 ms	3.118 ms    -31.9%
future/7                 5.523 ms	3.833 ms    -30.6%
multisigSM/1             3.373 ms	2.140 ms    -36.6%
multisigSM/2             3.403 ms	2.151 ms    -36.8%
multisigSM/3             3.456 ms	2.197 ms    -36.4%
multisigSM/4             3.484 ms	2.242 ms    -35.6%
multisigSM/5             3.977 ms	2.733 ms    -31.3%
multisigSM/6             3.360 ms	2.152 ms    -36.0%
multisigSM/7             3.394 ms	2.165 ms    -36.2%
multisigSM/8             3.445 ms	2.203 ms    -36.1%
multisigSM/9             3.471 ms	2.242 ms    -35.4%
multisigSM/10            3.995 ms	2.727 ms    -31.7%
vesting/1                2.904 ms	2.099 ms    -27.7%
vesting/2                2.600 ms	1.830 ms    -29.6%
vesting/3                2.901 ms	2.097 ms    -27.7%
marlowe/trustfund/1      9.293 ms	5.940 ms    -36.1%
marlowe/trustfund/2      7.418 ms	4.265 ms    -42.5%
marlowe/zerocoupon/1     9.478 ms	6.127 ms    -35.4%
marlowe/zerocoupon/2     6.813 ms	3.753 ms    -44.9%
```

Eh?  I'm not sure that I believe that.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs [hunk 1]
```diff
@@ -50,9 +48,6 @@ abstractly specifiable. It's an implementation detail.
 -}
 
 type Plain f (uni :: GHC.Type -> GHC.Type) (fun :: GHC.Type) = f TyName Name uni fun ()
--- | Caches Memory usage for builtin costing
--- | NOT the amount of memory it cost to calculate this value.
-type WithMemory f (uni :: GHC.Type -> GHC.Type) (fun :: GHC.Type) = f TyName Name uni fun ExMemory
```
Comment (kwxm) [inline plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs:55]: I was tempted to remove `Plain` as well, but that's used in some of the tests.

**Comment by michaelpj** [comment]
Argh! I knew this fix was buried in a branch of Simon's, but somehow I thought we'd dealt with it already.

Possibly I inflated your `master` run by also running something at the same time, sorry...

**Comment by kwxm** [comment]
> Possibly I inflated your `master` run by also running something at the same time, sorry...

I don't think so.  I still have the figures from benchmarking Roman's STvars branch last night and they seem to be pretty much the same as the results for master (see https://github.com/input-output-hk/plutus/pull/3012#issuecomment-820766586).  I'll do it again later though, and try nofib too.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs (hunk 1 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs:55]: I wouldn't mind expanding that type synonym at the call site and removing it here. But I don't care much either way.

**Review by effectfully — APPROVED** [review]
It seems unlikely that these changes can result in such a huge speedup given that the numbers are still within `Int64`... And the results are not particularly consistent (but not too inconsistent either). Maybe try rerunning the benchmarks? Or we can just not care.

**Comment by kwxm** [comment]
Please don't merge this just yet: I'll do that this evening.  I'm seeing similar speedup on my laptop, so I want to check the nofib benchmarks as well.

**Comment by kwxm** [comment]
I re-ran the validation benchmarks on both the server and my laptop and got similar reductions to the earlier ones.  I also ran the nofib benchmarks and got a uniform reduction, but nowhere near as dramatic.  Maybe we've reduced some constant overhead?

```
      	      	      	  master      	  new
---------------------------------------------------------
clausify/formula1        133.8 ms       126.5 ms    -5.5%
clausify/formula2        165.5 ms       156.4 ms    -5.5%
clausify/formula3        457.4 ms       435.9 ms    -4.7%
clausify/formula4        622.8 ms       594.3 ms    -4.6%
clausify/formula5        2.982 s        2.833 s     -5.0%
knights/4x4              317.0 ms       303.8 ms    -4.2%
knights/6x6              935.4 ms       894.4 ms    -4.4%
knights/8x8              1.581 s        1.515 s     -4.2%
primetest/05digits       186.5 ms       181.9 ms    -2.5%
primetest/08digits       339.8 ms       333.4 ms    -1.9%
primetest/10digits       476.2 ms       466.3 ms    -2.1%
primetest/20digits       952.0 ms       930.8 ms    -2.2%
primetest/30digits       1.380 s        1.349 s     -2.2%
primetest/40digits       1.844 s        1.803 s     -2.2%
primetest/50digits       1.812 s        1.774 s     -2.1%
queens4x4/bt             54.20 ms       51.63 ms    -4.7%
queens4x4/bm             69.79 ms       67.19 ms    -3.7%
queens4x4/bjbt1          66.85 ms       64.13 ms    -4.1%
queens4x4/bjbt2          71.69 ms       68.63 ms    -4.3%
queens4x4/fc             171.7 ms       163.8 ms    -4.6%
queens5x5/bt             726.3 ms       698.8 ms    -3.8%
queens5x5/bm             793.2 ms       762.7 ms    -3.8%
queens5x5/bjbt1          854.4 ms       819.9 ms    -4.0%
queens5x5/bjbt2          904.5 ms       867.5 ms    -4.1%
queens5x5/fc             2.185 s        2.097 s     -4.0%
```

**Comment by michaelpj** [comment]
Seems plausible to me!

**Comment by kwxm** [comment]
I wondered if the speedup was because the original `withMemory` function was in a separate module, so I put

```runCek runtime mode emitting term =
    runCekM runtime mode emitting $ do
        spendBudget BAST (ExBudget 0 (termAnn memTerm))
        computeCek [] mempty memTerm
  where
    memTerm = (memoryUsage term) <$  term   -- or even  1 <$ term
```

and that didn't have any effect.  When I replaced that with

```
 where
    memTerm = withMemory term
    withMemory =
        \case
         Constant () v      -> Constant (memoryUsage v) v
         Builtin  () b      -> Builtin  1 b
         ...
```
the speedup reappeared.  Maybe the explicit traversal is quicker for some reason, but I find it surprising that it makes such a difference.

**Comment by kwxm** [comment]
Aha!  I bet this'll speed up even more if we stop using the annotations for memory usage so that we can ditch the traversal altogether.

**Pull request merged**